.github/workflows/turing.yaml

```diff
@@ -236,8 +236,7 @@ jobs:
         with:
           path: api
           key: |
-            ${{ runner.os }}-vendor-${{ hashFiles('api/go.mod') }}
```

Comment:
Also, I'm having trouble identifying what the hash corresponds to. I tried stripping down the [hash logic](https://github.com/actions/runner/blob/main/src/Misc/expressionFunc/hashFiles/src/hashFiles.ts) here but the hex doesn't match:

https://codesandbox.io/s/recursing-raman-emy7f?file=/index.js

Reply:
I don't understand how the caching works. It builds in the container right? Does it pull the vendor outside the container then builds it in the container?

Reply:
try this? https://github.community/t/hashfiles-with-multiple-patterns/17168/3 hash the router and experiment folders as well

Reply:
I think all you need to do is as suggested by @ashwinath, to modify`key` to include hashes of internal `api`'s dependencies (engines/experiment and engines/router). Something like
`${{ runner.os }}-vendor-${{ hashFiles('api/go.mod') }}-${{ hashFiles('engines/experiment/go.mod') }}-${{ hashFiles('engines/router/go.mod') }}`

Reply:
As regarding why do we need `restore-keys`, the logic of it is well described in the docs (https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key), but to put it into context, let me illustrate this by this example:
1. You have your application with v1 of dependencies. After successfully running the pipeline for the first time, it will create a cache for your dependencies under the `ubuntu-vendor-<v1-god.mod-hash>`.
2. Then, you updated your `go.mod` to add/update/delete new dependencies, so your workflows will they to look for the cache ander `ubuntu-vendor-<v2-god.mod-hash>` which doesn't exist yet.
3. If you have your `restore-keys` set to `ubuntu-vendor-`, then it partially matches the key of `ubuntu-vendor-<v1-god.mod-hash>`, hence it will use this cache to speed-up your build.
4. Since `ubuntu-vendor-<v1-god.mod-hash>` does not 100% match your dependencies set, you still need to run `go mod vendor` to fetch/update missing dependencies

It all works well with properly versioned public dependencies, however, there could be some quirks if we use `replace` to point experiment/router dependencies into the private repo. I don't know how exactly it works, so some investigation might be required. But maybe if you update versions of experiment/router (it's currently set to `v0.0.0`) to something else, then `go mod vendor` will try to re-fetch it and will update it in the previously fetched cache from the v1 cache archive ()

Reply:
>I  think all you need to do is as suggested by @ashwinath, to modify key to include hashes of internal api's dependencies (engines/experiment and engines/router). Something like
${{ runner.os }}-vendor-${{ hashFiles('api/go.mod') }}-${{ hashFiles('engines/experiment/go.mod') }}-${{ hashFiles('engines/router/go.mod') }}

I'm afraid the `go.mod` is not enough. A change in logic in the `engines/experiment` module will not affect its `go.mod` file (or `go.sum`, for that matter) - adding them to the hash pattern will not detect that something changed in this scenario and this actually the case in this PR. But I think @ashwinath's idea of "hash the router and experiment **folders** " may do the trick, but (from what I understand) not if we use `restore-keys` (below).

Thanks for the detailed explanation on `restore-keys` - it is much clearer to me. In step #4, it is expected that the outdated dependencies are re-updated. So, regardless of what cached vendors have been restored, it is supposed to correct itself in the `Build Docker image` step since `make build-image` also does a `go mod vendor`. But, as you've pointed out, this fails to update the local dependency - this is strange to me, I've never encountered this in my local development. This is why I'm not sure if the `build-api` step's cache pattern needs to be changed at all:
* Using the deps' `go.mod` is not exhaustive
* Hashing the deps' folders as Ashwin suggested may be good but it will create a new cache key almost always which can easily hit the max cache size supported by Github.

I will perhaps try to exclude `github.com/gojek/turing` path from the cache

Reply:
Eh wait: https://github.com/gojek/turing/runs/4208336794?check_suite_focus=true#step:4:844

The problem is that the `vendor` data is actually updated but the API code is older. I think it's because we are caching `api` instead of `api/vendor`: https://github.com/gojek/turing/blob/main/.github/workflows/turing.yaml#L238

Reply:
Ok, that was it.

Reply:
> I think all you need to do is as suggested by @ashwinath, to modify`key` to include hashes of internal `api`'s dependencies (engines/experiment and engines/router). Something like `${{ runner.os }}-vendor-${{ hashFiles('api/go.mod') }}-${{ hashFiles('engines/experiment/go.mod') }}-${{ hashFiles('engines/router/go.mod') }}`

I missed this, I think go.mod isn't enough, a digest should be generated based on all the source code as well. I don't think it is an extremely slow process but we can test it out.